packages/astro/src/core/build/plugins/plugin-ssr.ts

```diff
@@ -73,12 +69,224 @@ function vitePluginSSR(
+		},
+		async generateBundle(_opts, bundle) {
+			// Add assets from this SSR chunk as well.
+			for (const [_chunkName, chunk] of Object.entries(bundle)) {
+				if (chunk.type === 'asset') {
+					internals.staticFiles.add(chunk.fileName);
+				}
+			}
+
+			for (const [chunkName, chunk] of Object.entries(bundle)) {
+				if (chunk.type === 'asset') {
+					continue;
+				}
+				if (chunk.modules[RESOLVED_SSR_VIRTUAL_MODULE_ID]) {
+					internals.ssrEntryChunk = chunk;
+					delete bundle[chunkName];
+				}
+			}
+		},
+	};
+}
+
+export function pluginSSRServer(
+	options: StaticBuildOptions,
+	internals: BuildInternals
+): AstroBuildPlugin {
+	const ssr =
+		options.settings.config.output === 'server' || isHybridOutput(options.settings.config);
+	return {
+		build: 'ssr',
+		hooks: {
+			'build:before': () => {
+				let vitePlugin =
+					// config.build object is optional, so we check NOT EQUAL against "serverless" instead
+					ssr && options.settings.config.build?.mode !== 'serverless'
+						? vitePluginSSRServer(internals, options.settings.adapter!, options)
+						: undefined;
+
+				return {
+					enforce: 'after-user-plugins',
+					vitePlugin,
+				};
+			},
+			'build:post': async ({ mutate }) => {
+				if (!ssr) {
+					return;
+				}
+
+				if (options.settings.config.build?.mode === 'serverless') {
+					return;
+				}
+
+				if (!internals.ssrEntryChunk) {
+					throw new Error(`Did not generate an entry chunk for SSR`);
+				}
+				// Mutate the filename
+				internals.ssrEntryChunk.fileName = options.settings.config.build.serverEntry;
+
+				const code = await injectManifest(options, internals, internals.ssrEntryChunk);
+				mutate(internals.ssrEntryChunk, 'server', code);
+			},
+		},
+	};
+}
+
+export const SERVERLESS_MODULE_ID = '@astro-page-serverless:';
+export const RESOLVED_SERVERLESS_MODULE_ID = '\0@astro-page-serverless:';
+
+function vitePluginSSRServerless(
+	internals: BuildInternals,
+	adapter: AstroAdapter,
+	options: StaticBuildOptions
+): VitePlugin {
+	return {
+		name: '@astrojs/vite-plugin-astro-ssr-serverless',
+		enforce: 'post',
+		options(opts) {
+			if (options.settings.config.build?.mode === 'serverless') {
+				const inputs: Set<string> = new Set();
+
+				for (const path of Object.keys(options.allPages)) {
+					inputs.add(getVirtualModulePageNameFromPath(SERVERLESS_MODULE_ID, path));
+				}
+
+				return addRollupInput(opts, Array.from(inputs));
+			}
+		},
+		resolveId(id) {
+			if (id.startsWith(SERVERLESS_MODULE_ID)) {
+				return '\0' + id;
+			}
+		},
+		async load(id) {
+			if (id.startsWith(RESOLVED_SERVERLESS_MODULE_ID)) {
+				const {
+					settings: { config },
+					allPages,
+				} = options;
+				const imports: string[] = [];
+				const contents: string[] = [];
+				const exports: string[] = [];
+
+				const path = getPathFromVirtualModulePageName(RESOLVED_SERVERLESS_MODULE_ID, id);
+				const virtualModuleName = getVirtualModulePageNameFromPath(ASTRO_PAGE_MODULE_ID, path);
```

Comment:
Here we juggle from `\0@astro-serverless-page:src/pages/index@_@astro` to `@astro-page:src/pages/index@_@astro`